### DISCUSSION TITLE
[BREAKING CHANGE] New Top Language Detection Method

### DISCUSSION THREAD

**Discussion by anuraghazra** [anuraghazra/github-readme-stats#1027] [Ideas]
As you all might know there are various bugs/issues regarding the top languages calculation.

### The problem

The main issue i see is that people often get confused by how the calculations are done. 

Currently the top languages are calculated based on **how much** code in bytes you have in a particular language and then we choose the top languages. 
This method is the main reason people are confused about the calculations, because normally users perceive how much they code in languages by **how many** repositories they have with that particular language.

#### Quirks with the current calculation method

- Confusion about the calculations. (https://github.com/anuraghazra/github-readme-stats/issues/136#issuecomment-665172181)
- Repositories might have vendor code or auto generated code which would make the calculations wrong. (#153)
- If some language have exaggerated code bytes then it becomes the dominant language. (https://github.com/anuraghazra/github-readme-stats/issues/358#issuecomment-673327755)
- Users aren't satisfied with the method.

### The Solution

The most straight forward solution I see is that instead of calculating **how much** code they have, we can calculate **how many**  repositories they have with the languages.

##### Related issues
#432 #403 #270 #136 #358

**Comment by saurabhdaware** [comment]
Not sure but the current methods seems right to me. Imagine having 2 HTML, JS repositories with more 51% HTML in both.

According to the new proposed method, I wouldn't know JavaScript.

So in my opinion, the current method makes more sense.

**Comment by anuraghazra** [comment]
@saurabhdaware We would not just take the primary language of the individual repo, we would also calculate top 10 langs of individual repos too. 

That is we are already doing :-

https://github.com/anuraghazra/github-readme-stats/blob/6e73a0035dfebf7411f4763b7ce49df2f8737e36/src/fetchers/top-languages-fetcher.js#L14

So it would look like -:

HTML ---------- some%
Javascript ------ some%

**Comment by saurabhdaware** [comment]
I am not sure if I understood. Wouldn't calculating top 10 languages of each repository same as calculating how much code the user has in bytes? that's how GitHub calculates the percentage as well no?

**Comment by anuraghazra** [comment]
No we would just "count" them and in current method we get the "language.size" reduce and sum it up and then sort it.

**Comment by saurabhdaware** [comment]
Oh ok so in this example
> Imagine having 2 HTML, JS repositories with more 51% HTML in both.

I would have 
51% HTML 
49% JavaScript 
right?

**Comment by anuraghazra** [comment]
Repo 1 - JS x1 & HTML x1
Repo 2 - JS x1 & HTML x1

HTML - 50%
JS - 50%


We would just *count* them.

**Comment by saurabhdaware** [comment]
Oh cool. Seems good to me then.

**Comment by DenverCoder1** [comment]
I think I like this new way better. I did one very large C# project earlier this year and now it says C# is my top language (over 52%) even though it is definitely _not_ what I do the most of.

**Comment by DenverCoder1** [comment]
If there are people that are more happy with the current way, there could possibly be an additional parameter that will switch between the different calculation modes?

**Comment by anuraghazra** [comment]
> If there are people that are more happy with the current way, there could possibly be an additional parameter that will switch between the different calculation modes?

Unfortunately we cannot do that, it would make the logic complex & we would have two different statistics. it would hamper the consistency.

**Comment by anuraghazra** [comment]
I will firstly publish experimental query param to enable this and then if people likes it i would make it default.

**Comment by Bas950** [comment]
I personally wouldnt use this new one, I think the current one is better.

> I think I like this new way better. I did one very large C# project earlier this year and now it says C# is my top language (over 52%) even though it is definitely _not_ what I do the most of.

You could exclude the language, but in my PR (https://github.com/anuraghazra/github-readme-stats/pull/480) I am making it so you can just exclude a repo. Which is probs better.

**Comment by anuraghazra** [comment]
@Bas950 yeah I can understand what you are saying, but the main reason why soo many people uses github-readme-stats is because of it's simplicity and ease of use. Of course we can add "exclude_repo" options and make it better but the thing is that not many people have the time/patience to go through all of their repositories and check which one has some vendor code and exclude them one by one, not to mention this is impractical for users who have lot of repos.

So this is why i'm considering this new approach which would mitigate these issues.

**Comment by crazy-max** [comment]
@anuraghazra Check https://github.com/github/linguist

**Reply by cicirello** [reply]
@anuraghazra definitely check out linguist. Some of the issues with language stats can be handled with no changes to github-readme-stats itself. For example, Linguist automatically attempts to exclude vendored code, documentation, and some other stuff from what is reported for a repo. If it gets it wrong, either including something that should be included, or the opposite, you can configure linguist for that repo in a .gitattributes. Updating the docs for github-readme-stats to explain this and link to relevant part of linguist docs will partially cover what users want.

**Comment by mchelen-gov** [comment]
> #### Quirks with the current calculation method
> * Confusion about the calculations. ([#136 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/136#issuecomment-665172181))
> * Repositories might have vendor code or auto generated code which would make the calculations wrong. (#153)
> * If some language have exaggerated code bytes then it becomes the dominant language. ([#358 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/358#issuecomment-673327755))
> * Users aren't satisfied with the method.

@anuraghazra Does the current method only look at repos owned by the user or does it include other repos the user has contributed to?

**Comment by ghost** [comment]
Consider i have five repositories.

1) C++ 50k line Machine Learning Library i wrote.
2) html page with hello
3) html page with hello
4) html page with hello
5) html page with hello 


Now repo approcah should give my top language as HTML! right??

**Comment by anuraghazra** [comment]
> Consider i have five repositories.
> 
> 1. C++ 50k line Machine Learning Library i wrote.
> 2. html page with hello
> 3. html page with hello
> 4. html page with hello
> 5. html page with hello
> 
> Now repo approcah should give my top language as HTML! right??

![](https://media1.tenor.com/images/7d838b33d845511d4b9876d13bc8762c/tenor.gif)

**Comment by Bas950** [comment]
> > #### Quirks with the current calculation method
> > 
> > * Confusion about the calculations. ([#136 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/136#issuecomment-665172181))
> > * Repositories might have vendor code or auto generated code which would make the calculations wrong. (#153)
> > * If some language have exaggerated code bytes then it becomes the dominant language. ([#358 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/358#issuecomment-673327755))
> > * Users aren't satisfied with the method.
> 
> @anuraghazra Does the current method only look at repos owned by the user or does it include other repos the user has contributed to?

I will be add a feature soon in PR tho that will allow you to opt-in to forks or opt-out of using forks, depends of what @anuraghazra wants to use.

**Comment by casperiv0** [comment]
Kinda off-topic question: Does the stats-card calculate private repos aswell? Thankyou.

**Comment by Vicellken** [comment]
> Kinda off-topic question: Does the stats-card calculate private repos aswell? Thankyou.

looking forward to this feature as well!

**Comment by benstigsen** [comment]
Will this make it support other languages, or should I make a seperate PR for this? Currently I can't see GDScript on my stats.

**Comment by ghost** [comment]
Don't think so.

**Comment by jcubic** [comment]
Please check issue #450 where I've provied GraphQL (I don't remember if it works and if test it) query that show most stared repos, alternative maybe to get repos with most commits (but there are no order by number of commits yet).

Using default 100 repos is stupid because the order can be random and those repos can have code that you didn't added single commit and get the code from somewhere else. Not only forks have forked code, if code is not on GitHub you can't fork you need to copy the code into your own repo.

**Comment by Potherca** [comment]
> Please check issue #450 where I've provied GraphQL query

I think the suggestion (in that ticket) to use `orderBy` might be a good one. The next issue will, of course, be "Sort by what?", which will undoubtedly lead to "I want to sort by X, not Y, can you make it configurable?". But I think the basic premise is a good addition to resolving this rather sticky puzzle.

> (I don't remember if it works and if test it)

I just ran it through the [graphql explorer](https://docs.github.com/en/graphql/overview/explorer), it works :+1: 

> those repos can have code that you didn't added single commit and get the code from somewhere else. Not only forks have forked code, if code is not on GitHub you can't fork you need to copy the code into your own repo.

For such repos, I would suggest using the `exclude_repo` setting (or just creating a separate org and moving such repos there).
I don't think we should really expect such a level of intelligence from a project such as this. I think the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) would apply here.

> Using default 100 repos is stupid

Not stupid.  _Easy_. The API won't let you get more in a single request, so multiple requests would need to be made.
Making more calls means more code, more work, potentially more issues, etc.

**Comment by jcubic** [comment]
By stupid I mean default 100 by sorting like this. Even 10 repos is better if the sorting is done right, most faved repos or repos with most commits maybe most recent commits. Anything but default order which looks like random with fixed seed.

**Comment by Potherca** [comment]
Thank you for clarifying. 🙌

> Even 10 repos is better if the sorting is done right

Yes! I **completely** agree with this!

I was thinking about this some more and I think the main issue here is conflicting use-cases... The GH API gives language stats for _Repositories_, not for _Users_. It might be easier to add a separate card and/or split the use-case to support both sides?

@anuraghazra If @jcubic and myself were to set up some examples (using various queries and parameters mentioned in this thread), would you be available to play around with them? We could ask some of the other people in this (an linked) issues for feedback as well.

It would be a shame to let all the hard work and thoughts that went into this issue come to nothing...

**Comment by anuraghazra** [comment]
> anuraghazra If jcubic and myself were to set up some examples (using various queries and parameters mentioned in this thread), would you be available to play around with them? We could ask some of the other people in this (an linked) issues for feedback as well.

@Potherca feel free to experiment with different ways to make it more accurate I can surely take a look at them and give some feedbacks on it.

**Comment by anuraghazra** [comment]
Another possible way to count language stats is by using github's search api https://docs.github.com/en/rest/reference/search#search-code

![image](https://user-images.githubusercontent.com/35374649/106381019-0606dc80-63dc-11eb-9749-d4f3d2e90df6.png)

**Comment by ghost** [comment]
see this api
https://codetabs.com/count-loc/count-loc-online.html

It could be helpful in taking into consideration total lines of codes of
specific repository and its language,

**Comment by jcubic** [comment]
it will not work for the case where somone fork repo that is not git fork, but copy that have single commit. I have one repo like this that is written in C and Lua and it will give me those languages that I've never written even single line of it.

**Comment by mushahidq** [comment]
> Repo 1 - JS x1 & HTML x1
> Repo 2 - JS x1 & HTML x1
> 
> HTML - 50%
> JS - 50%
> 
> We would just _count_ them.

Is the proposed solution, just to count the languages which appear most in the result of the GitHub API response?

**Comment by anuraghazra** [comment]
> > Repo 1 - JS x1 & HTML x1
> > Repo 2 - JS x1 & HTML x1
> > HTML - 50%
> > JS - 50%
> > We would just _count_ them.
> 
> Is the proposed solution, just to count the languages which appear most in the result of the GitHub API response?

No that sucks https://github.com/anuraghazra/github-readme-stats/issues/481#issuecomment-698852675

**Comment by andreped** [comment]
> > > Repo 1 - JS x1 & HTML x1
> > > Repo 2 - JS x1 & HTML x1
> > > HTML - 50%
> > > JS - 50%
> > > We would just _count_ them.
> > 
> > 
> > Is the proposed solution, just to count the languages which appear most in the result of the GitHub API response?
> 
> No that sucks [#481 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/481#issuecomment-698852675)

Is it that bad though? I guess the problem here is that we are unsure what we want these numbers to represent. I wasn't expecting these numbers to represent the distribution of total number of lines I wrote using each language. That would also be a bad estimate as in C I write quite many more lines for a simple sum compared to in Python. Doesn't necessarily mean I do more C than Python. Again, it really depends what one want these estimates to measure.

At least for me doing multiple projects, it would be cool to have an estimate stating **repository-wise** what is the most common language you have used. Initially, this is actually what I thought these numbers represented. But there are scenarios where such a measure might be suboptimal, as aforementioned, if one assume otherwise.

I do not think there is an optimum here that suits all users. Perhaps it could be an option to support both designs, or even multiple? That would at least solve my issue, and thus make me happy :]

But perhaps having more than one design that estimate these measures might introduce even more noise into how to interpret these values... Idk anymore

**Comment by foxt** [comment]
It'd also be nice to be able to exclude forks.

My card
![my card](https://github-readme-stats.vercel.app/api/top-langs/?username=thelmgn&layout=compact)
Currently shows 86% Python because I'm making a minor PR to a Python repo.

**Comment by andreped** [comment]
I agree with @theLMGN . 

I have a fork on my repo, which I haven't contributed to yet, but which apparently contain a shit ton of C#, which I have never used. This results in my "Most used languages" to be roughly 80% C#, which is sort of funny considering I have contributed to roughly 40 open repos, of which are Python/C++.

I'm also wondering if C# and C++ are switched, or if C++ code is misinterpreted as C# in github-readme-stats. According to github-readme-stats I do not do any C++, but I have contributed to C++ projects (forks).

**Comment by andreped** [comment]
@theLMGN couldn't you just use the **exclude_repo** option to exclude that one repository? Since you only did a minor PR, including this repo in the calculating is probably not necessary?

**Comment by y9c** [comment]
Why no add a argument `&by=xxx` to resolve this paradox. 
xxx can be `repo`, `line`, `file`
